docs/rfcs/0012-introduce-ui-store.md

```diff
@@ -0,0 +1,161 @@
+
+* Simplicity for `Store`s that are meant for UI facing work.
+* Unit testing can now occur on the test framework's thread.
+* Fewer resources needed for context shifting between threads[^1].
+
+For an example of thread simplicity, an `Engine` typically has its own 'engine thread' to perform async work and post/request results to the main thread (these APIs are identified with the `@UiThread` annotation). Once we get the callback for those results, we then need to dispatch an action to the store that will then happen on a `Store` thread. Feature components then observe for state changes and then make UI changes on the main thread. A simplified form of this thread context switching can be seen in the example below:
+
+```kotlin
+// engine thread
+engineView.requestApiResult { result ->
+  // received on the main thread.
+  store.dispatch(UpdateResultAction(result))
+}
+
+// store thread
+fun reduce(state: State, action: Action) {
+  is UpdateResultAction -> {
+    // do things here.
+  }
+}
+
+// store thread
+Middleware {
+  override fun invoke(
+    context: MiddlewareContext<State, Action>,
+    next: (Action) -> Unit,
+    action: Action,
+  ) {
+    // perform side-effects that also happen on the store thread.
+  }
+}
+
+// main thread
+store.flowScoped { flow ->
+  flow.collect {
+    // perform work on the main thread.
+  }
+}
+```
+
+With the changes in this RFC, this switching of threads can be reduced (notable comments marked with 📝):
+
+```kotlin
+// engine thread
+engineView.requestApiResult { result ->
+  // received on the main thread.
+  store.dispatch(UpdateResultAction(result))
+}
+
+// 📝 main thread - now on the same thread, processed immediately.
+fun reduce(state: State, action: Action) {
+  is UpdateResultAction -> {
+    // do things here.
+  }
+}
+
+// 📝 main thread - now on the same thread, processed immediately.
+Middleware {
+  override fun invoke(
+    context: MiddlewareContext<State, Action>,
+    next: (Action) -> Unit,
+    action: Action,
+  ) {
+    // 📝 perform side-effects that now happen on the main thread.
+  }
+}
+
+// main thread
+store.flowScoped { flow ->
+  flow.collect {
+      // perform work on the main thread.
+    }
+  }
+}
+```
+
+Additionally, from [performance investigations already done][2], we know that Fenix creates over a hundred threads within a few seconds of startup. Reducing the number of threads for Stores that do not have a strong requirement to run on a separate thread will lower the applications memory footprint.
+
+## Guide-level explanation
+
+Extending the existing `Store` class to use the `Dispatchers.Main.immediate` will ensure that UI stores will stay on the same UI thread and have that work done immediately. Using a distinct class named `UiStore` also makes it clear to the developer that this is work that will be done on the UI thread and its implications will be made a bit more clear when it's used.
+
+```kotlin
+@MainThread
+open class UiStore<S : State, A : Action>(
```

Comment:
One other small implementation detail note while I'm here:

My preference would be for subtypes of UiStore to be typealiases, and leave `UiStore` as a closed class. This is again because I think inheritance makes code harder to trace and easier to introduce issues. For example, I think this is equivalent to the use cases you're imagining:

```
class UiStore<S : State, A : Action>(...) : Store<S, A>

typealias FragmentStore = UiStore<FragmentState, FragmentAction>
```